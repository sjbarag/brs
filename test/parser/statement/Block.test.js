const brs = require("brs");
const { Lexeme } = brs.lexer;
const { BrsString, Int32 } = brs.types;

const { token, identifier, EOF } = require("../ParserTests");

describe("block statements", () => {
    it("includes surrounding whitespace for function body", () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            sub Main()
                age = 1
            end sub
        `);
        const { statements, errors } = brs.parser.Parser.parse(tokens);
        expect(errors.length).toEqual(0);

        let blockLocation = statements[0].func.body.location;
        //the block location should begin immediately following `sub Main()`, and end immediately preceeding the first character of `end sub`
        expect(blockLocation).toEqual({
            file: "",
            start: { line: 2, column: 22 },
            end: { line: 4, column: 12 },
        });

        expect(statements).toMatchSnapshot();
    });

    it("includes surrounding whitespace for if-elseif-else statements", () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            sub Main()
                if true then
                    print "true"
                else if true then
                    print "true"
                else
                    print "true"
                end if
            end sub
        `);
        const { statements, errors } = brs.parser.Parser.parse(tokens);
        expect(errors.length).toEqual(0);

        const ifStatement = statements[0].func.body.statements[0];

        const thenBlock = ifStatement.thenBranch;
        expect(thenBlock.location).toEqual({
            file: "",
            start: { line: 3, column: 28 },
            end: { line: 5, column: 16 },
        });

        const elseIfBlock = ifStatement.elseIfs[0].thenBranch;
        expect(elseIfBlock.location).toEqual({
            file: "",
            start: { line: 5, column: 33 },
            end: { line: 7, column: 16 },
        });

        const elseBlock = ifStatement.elseBranch;
        expect(elseBlock.location).toEqual({
            file: "",
            start: { line: 7, column: 20 },
            end: { line: 9, column: 16 },
        });

        expect(statements).toMatchSnapshot();
    });

    it("includes surrounding whitespace for a for-loop body", () => {
        const { tokens } = brs.lexer.Lexer.scan(`
            sub Main()
                For i = 0 To 10 Step 1
                    print i
                End For
            end sub
        `);
        const { statements, errors } = brs.parser.Parser.parse(tokens);
        expect(errors.length).toEqual(0);

        let loopBlock = statements[0].func.body.statements[0].body;
        //the block location should begin immediately following `step 1`, and end immediately preceeding the first character of `end for`
        expect(loopBlock.location).toEqual({
            file: "",
            start: { line: 3, column: 38 },
            end: { line: 5, column: 16 },
        });

        expect(statements).toMatchSnapshot();
    });
});
