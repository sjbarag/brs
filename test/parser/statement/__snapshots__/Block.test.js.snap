// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`block statements includes surrounding whitespace for a for-loop body 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 6,
          },
          "file": "",
          "start": Object {
            "column": 22,
            "line": 2,
          },
        },
        "statements": Array [
          For {
            "body": Block {
              "location": Object {
                "end": Object {
                  "column": 16,
                  "line": 5,
                },
                "file": "",
                "start": Object {
                  "column": 38,
                  "line": 3,
                },
              },
              "statements": Array [
                Print {
                  "expressions": Array [
                    Variable {
                      "name": Object {
                        "isReserved": false,
                        "kind": "Identifier",
                        "literal": undefined,
                        "location": Object {
                          "end": Object {
                            "column": 27,
                            "line": 4,
                          },
                          "file": "",
                          "start": Object {
                            "column": 26,
                            "line": 4,
                          },
                        },
                        "text": "i",
                      },
                      "type": "Variable",
                    },
                  ],
                  "tokens": Object {
                    "print": Object {
                      "isReserved": true,
                      "kind": "Print",
                      "literal": undefined,
                      "location": Object {
                        "end": Object {
                          "column": 25,
                          "line": 4,
                        },
                        "file": "",
                        "start": Object {
                          "column": 20,
                          "line": 4,
                        },
                      },
                      "text": "print",
                    },
                  },
                  "type": "Print",
                },
              ],
              "type": "Block",
            },
            "counterDeclaration": Assignment {
              "name": Object {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 21,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 20,
                    "line": 3,
                  },
                },
                "text": "i",
              },
              "tokens": Object {
                "equals": Object {
                  "isReserved": false,
                  "kind": "Equal",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 23,
                      "line": 3,
                    },
                    "file": "",
                    "start": Object {
                      "column": 22,
                      "line": 3,
                    },
                  },
                  "text": "=",
                },
              },
              "type": "Assignment",
              "value": Literal {
                "_location": Object {
                  "end": Object {
                    "column": 25,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 24,
                    "line": 3,
                  },
                },
                "type": "Literal",
                "value": Int32 {
                  "kind": 4,
                  "value": 0,
                },
              },
            },
            "finalValue": Literal {
              "_location": Object {
                "end": Object {
                  "column": 31,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 29,
                  "line": 3,
                },
              },
              "type": "Literal",
              "value": Int32 {
                "kind": 4,
                "value": 10,
              },
            },
            "increment": Literal {
              "_location": Object {
                "end": Object {
                  "column": 38,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 37,
                  "line": 3,
                },
              },
              "type": "Literal",
              "value": Int32 {
                "kind": 4,
                "value": 1,
              },
            },
            "tokens": Object {
              "endFor": Object {
                "isReserved": false,
                "kind": "EndFor",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 23,
                    "line": 5,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 5,
                  },
                },
                "text": "End For",
              },
              "for": Object {
                "isReserved": true,
                "kind": "For",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 19,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 3,
                  },
                },
                "text": "For",
              },
              "step": Object {
                "isReserved": true,
                "kind": "Step",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 36,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 32,
                    "line": 3,
                  },
                },
                "text": "Step",
              },
              "to": Object {
                "isReserved": true,
                "kind": "To",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 28,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 26,
                    "line": 3,
                  },
                },
                "text": "To",
              },
            },
            "type": "For",
          },
        ],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 6,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 6,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 20,
          "line": 2,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 2,
        },
      },
      "text": "Main",
    },
    "type": "Function",
  },
]
`;

exports[`block statements includes surrounding whitespace for function body 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 4,
          },
          "file": "",
          "start": Object {
            "column": 22,
            "line": 2,
          },
        },
        "statements": Array [
          Assignment {
            "name": Object {
              "isReserved": false,
              "kind": "Identifier",
              "literal": undefined,
              "location": Object {
                "end": Object {
                  "column": 19,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 16,
                  "line": 3,
                },
              },
              "text": "age",
            },
            "tokens": Object {
              "equals": Object {
                "isReserved": false,
                "kind": "Equal",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 21,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 20,
                    "line": 3,
                  },
                },
                "text": "=",
              },
            },
            "type": "Assignment",
            "value": Literal {
              "_location": Object {
                "end": Object {
                  "column": 23,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 22,
                  "line": 3,
                },
              },
              "type": "Literal",
              "value": Int32 {
                "kind": 4,
                "value": 1,
              },
            },
          },
        ],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 4,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 4,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 20,
          "line": 2,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 2,
        },
      },
      "text": "Main",
    },
    "type": "Function",
  },
]
`;

exports[`block statements includes surrounding whitespace for if-elseif-else statements 1`] = `
Array [
  Function {
    "func": Function {
      "body": Block {
        "location": Object {
          "end": Object {
            "column": 12,
            "line": 10,
          },
          "file": "",
          "start": Object {
            "column": 22,
            "line": 2,
          },
        },
        "statements": Array [
          If {
            "condition": Literal {
              "_location": Object {
                "end": Object {
                  "column": 23,
                  "line": 3,
                },
                "file": "",
                "start": Object {
                  "column": 19,
                  "line": 3,
                },
              },
              "type": "Literal",
              "value": BrsBoolean {
                "kind": 2,
                "value": true,
              },
            },
            "elseBranch": Block {
              "location": Object {
                "end": Object {
                  "column": 16,
                  "line": 9,
                },
                "file": "",
                "start": Object {
                  "column": 20,
                  "line": 7,
                },
              },
              "statements": Array [
                Print {
                  "expressions": Array [
                    Literal {
                      "_location": Object {
                        "end": Object {
                          "column": 32,
                          "line": 8,
                        },
                        "file": "",
                        "start": Object {
                          "column": 26,
                          "line": 8,
                        },
                      },
                      "type": "Literal",
                      "value": BrsString {
                        "kind": 3,
                        "value": "true",
                      },
                    },
                  ],
                  "tokens": Object {
                    "print": Object {
                      "isReserved": true,
                      "kind": "Print",
                      "literal": undefined,
                      "location": Object {
                        "end": Object {
                          "column": 25,
                          "line": 8,
                        },
                        "file": "",
                        "start": Object {
                          "column": 20,
                          "line": 8,
                        },
                      },
                      "text": "print",
                    },
                  },
                  "type": "Print",
                },
              ],
              "type": "Block",
            },
            "elseIfs": Array [
              Object {
                "condition": Literal {
                  "_location": Object {
                    "end": Object {
                      "column": 28,
                      "line": 5,
                    },
                    "file": "",
                    "start": Object {
                      "column": 24,
                      "line": 5,
                    },
                  },
                  "type": "Literal",
                  "value": BrsBoolean {
                    "kind": 2,
                    "value": true,
                  },
                },
                "thenBranch": Block {
                  "location": Object {
                    "end": Object {
                      "column": 16,
                      "line": 7,
                    },
                    "file": "",
                    "start": Object {
                      "column": 33,
                      "line": 5,
                    },
                  },
                  "statements": Array [
                    Print {
                      "expressions": Array [
                        Literal {
                          "_location": Object {
                            "end": Object {
                              "column": 32,
                              "line": 6,
                            },
                            "file": "",
                            "start": Object {
                              "column": 26,
                              "line": 6,
                            },
                          },
                          "type": "Literal",
                          "value": BrsString {
                            "kind": 3,
                            "value": "true",
                          },
                        },
                      ],
                      "tokens": Object {
                        "print": Object {
                          "isReserved": true,
                          "kind": "Print",
                          "literal": undefined,
                          "location": Object {
                            "end": Object {
                              "column": 25,
                              "line": 6,
                            },
                            "file": "",
                            "start": Object {
                              "column": 20,
                              "line": 6,
                            },
                          },
                          "text": "print",
                        },
                      },
                      "type": "Print",
                    },
                  ],
                  "type": "Block",
                },
                "type": "ElseIf",
              },
            ],
            "thenBranch": Block {
              "location": Object {
                "end": Object {
                  "column": 16,
                  "line": 5,
                },
                "file": "",
                "start": Object {
                  "column": 28,
                  "line": 3,
                },
              },
              "statements": Array [
                Print {
                  "expressions": Array [
                    Literal {
                      "_location": Object {
                        "end": Object {
                          "column": 32,
                          "line": 4,
                        },
                        "file": "",
                        "start": Object {
                          "column": 26,
                          "line": 4,
                        },
                      },
                      "type": "Literal",
                      "value": BrsString {
                        "kind": 3,
                        "value": "true",
                      },
                    },
                  ],
                  "tokens": Object {
                    "print": Object {
                      "isReserved": true,
                      "kind": "Print",
                      "literal": undefined,
                      "location": Object {
                        "end": Object {
                          "column": 25,
                          "line": 4,
                        },
                        "file": "",
                        "start": Object {
                          "column": 20,
                          "line": 4,
                        },
                      },
                      "text": "print",
                    },
                  },
                  "type": "Print",
                },
              ],
              "type": "Block",
            },
            "tokens": Object {
              "else": Object {
                "isReserved": true,
                "kind": "Else",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 20,
                    "line": 7,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 7,
                  },
                },
                "text": "else",
              },
              "elseIfs": Array [
                Object {
                  "isReserved": false,
                  "kind": "ElseIf",
                  "literal": undefined,
                  "location": Object {
                    "end": Object {
                      "column": 23,
                      "line": 5,
                    },
                    "file": "",
                    "start": Object {
                      "column": 16,
                      "line": 5,
                    },
                  },
                  "text": "else if",
                },
              ],
              "endIf": Object {
                "isReserved": false,
                "kind": "EndIf",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 22,
                    "line": 9,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 9,
                  },
                },
                "text": "end if",
              },
              "if": Object {
                "isReserved": true,
                "kind": "If",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 18,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 16,
                    "line": 3,
                  },
                },
                "text": "if",
              },
              "then": Object {
                "isReserved": true,
                "kind": "Identifier",
                "literal": undefined,
                "location": Object {
                  "end": Object {
                    "column": 28,
                    "line": 3,
                  },
                  "file": "",
                  "start": Object {
                    "column": 24,
                    "line": 3,
                  },
                },
                "text": "then",
              },
            },
            "type": "If",
          },
        ],
        "type": "Block",
      },
      "end": Object {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 19,
            "line": 10,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 10,
          },
        },
        "text": "end sub",
      },
      "keyword": Object {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": 15,
            "line": 2,
          },
          "file": "",
          "start": Object {
            "column": 12,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": Array [],
      "returns": 11,
      "type": "Function",
    },
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": 20,
          "line": 2,
        },
        "file": "",
        "start": Object {
          "column": 16,
          "line": 2,
        },
      },
      "text": "Main",
    },
    "type": "Function",
  },
]
`;
